std/machines/shift_bb.asm

```diff
@@ -0,0 +1,70 @@
+use std::utils::unchanged_until;
+use std::utils::cross_product;
+use std::convert::int;
+
+// Shift for single bytes using an exhaustive table
+// TODO this way, we cannot prove anything that shifts by more than 31 bits.
+machine ByteShiftBB with
+    latch: latch,
+    operation_id: operation_id,
+    degree: 65536
+{
+    // P_C1 and P_C2 are both 16 bit limbs of P_C, where P_C1 is the less significant limb.
+    operation run<0> P_operation, P_A, P_B, P_ROW -> P_C1, P_C2;
+
+    col fixed latch = [1]*;
+    col fixed operation_id = [0]*;
+
+    let bit_counts = [256, 32, 4, 2];
+    let min_degree = std::array::product(bit_counts);
+    std::check::assert(std::prover::min_degree() >= std::array::product(bit_counts), || "The shift machine needs at least 65536 rows to work.");
+    let inputs = cross_product(bit_counts);
+    let a: int -> int = inputs[0];
+    let b: int -> int = inputs[1];
+    let row: int -> int = inputs[2];
+    let op: int -> int = inputs[3];
+    let P_A: col = a;
+    let P_B: col = b;
+    let P_ROW: col = row;
+    let P_operation: col = op;
+    let c: int -> int = |i| match op(i) {
+        0 => a(i) << (b(i) + (row(i) * 8)),
+        1 => (a(i) << (row(i) * 8)) >> b(i)
+    };
+    col fixed P_C1(i) { c(i) & 0xffff };
+    col fixed P_C2(i) { (c(i) >> 16) & 0xffff };
+}
+
+machine ShiftBB(byte_shift: ByteShiftBB) with
+    latch: latch,
+    operation_id: operation_id,
+    // Allow this machine to be connected via a permutation
+    call_selectors: sel,
+{
+    operation shl<0> A1, A2, B -> C1, C2;
```

Comment:
could call this `Ahi` and `Alow`, to make endianness clear

Reply:
Agree